### PR TITLE
reorganization of prompt generation by merging LinguisticPrograms

### PR DESCRIPTION
**Pull request by tobowers**
~This is a bit in progress (none of the gpt4 stuff is implemented), but I wanted to get some feedback before adding finishing details.~

This does not yet add things like persistence or working memory, but instead gives pretty high flexibility to anyone developing bots/souls and lets them cutomize.

Implemented Personality, ConversationCompressor, and PersonMemory using this and it's all pretty straight forward to implement.

### CODE REVIEW

File: core/src/programs/index.ts [hunk 1]
```diff
@@ -1 +1,17 @@
+import { ConversationProcessor } from "../conversationProcessor";
+import { Thought } from "../languageModels/memory";
+import { ProgramOutput } from "../linguisticProgramBuilder";
+
 export * from "./participationStrategies";
+
+/**
+ * Mental models are processors on a stream from a group of conversation messages (or none) that
```
Comment (kafischer) [inline core/src/programs/index.ts:8]:
```suggestion
 * Conversational Programs are processors on a stream from a group of conversation messages (or none) that
```

File: core/src/programs/index.ts [hunk 2]
```diff
@@ -1 +1,17 @@
...
+ * end up being compiled to a prompt for the LLM.
+ * They can manipulate the message history being sent to the LLM, and build up a context for the system message prompt.
+ */
+export interface ConversationalProgram {
+  update: (thoughts: Thought[], conversation: ConversationProcessor) => void;
```
Comment (kafischer) [inline core/src/programs/index.ts:13]: This is not for this PR, but I don't love this conversation dependency

File: core/src/programs/Ramble.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { rambleAction } from "../actions/ramble";
+import { ConversationalProgram } from "./index";
+
+export class RambleProgram implements ConversationalProgram {
```
Comment (kafischer) [inline core/src/programs/Ramble.ts:4]: great, this makes more sense now!

File: core/src/programs/Personality.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    ];
+
+    if (plan) {
+      content.unshift(plan);
+    }
+
+    return {
+      context: content,
+      output: outputs,
+      rememberances: [
```
Comment (kafischer) [inline core/src/programs/Personality.ts:54]: this is an aside (for later) i think we might want to change 'rememberances' to something like 'finalInstructions'. There's some more patterns that are more flexible than just the rememberances one, especially if we have something that guarantees output format

File: core/src/linguisticProgramBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    name: soul.blueprint.name,
+  };
+
+  const rememberance: ChatMessage = {
+    role: ChatMessageRoleEnum.System,
+    content: `
+
+${programOutputs.flatMap((p) => p.rememberances || []).join("\n")}
+
+Now, think through ${
```
Comment (kafischer) [inline core/src/linguisticProgramBuilder.ts:144]: this is an unrelated minor point - i don't know why the linter insists on these breaks. Makes these hard to read (I know I'e added most of them)

File: core/src/linguisticProgramBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+export interface OutputTag extends ContextTag {
+  optional?: boolean;
+}
+
+export type MarkedMessage = ChatMessage & { creator?: string };
+
+export interface ProgramOutput {
+  context: ContextTag[];
+  output: OutputTag[];
+  rememberances: string[];
```
Comment (kafischer) [inline core/src/linguisticProgramBuilder.ts:19]: i want to figure out how to remove this at some point. i have some ideas. just commenting for now

File: core/src/linguisticProgramBuilder.ts (hunk 5 above)
Comment (tobowers) [reply core/src/linguisticProgramBuilder.ts:144]: I 100% agree, I was just following the guidelines :)

**Comment by tobowers** [comment]
IMO this is good to go now (removed the WIP marker)
